File: LeetCode_Java/dp/LongestPalindromicSubstring.java
package dp;

public class LongestPalindromicSubstring {

    public String longestPalindrome(String s) {
        if (s.length() <= 1) {
            return s;
        }

        int n = s.length();
        boolean[][] dp = new boolean[n][n];
        int start = 0;
        int maxLength = 0;

        for (int i = n - 1; i >= 0; i--) {
            for (int j = i; j < n; j++) {
                if (s.charAt(j) == s.charAt(i) && (j - i <= 2 || dp[i + 1][j - 1])) {
                    dp[i][j] = true;
                    if (j - i + 1 > maxLength) {
                        start = i;
                        maxLength = j - i + 1;
                    }
                }
            }
        }

        return s.substring(start, start + maxLength);
    }

    public static void main(String[] args) {
        LongestPalindromicSubstring solution = new LongestPalindromicSubstring();
        String s = "babad";
        System.out.println(solution.longestPalindrome(s)); // Output: "bab" or "aba"
    }

//    public String longestPalindrome(String s) {
//        if (s.length() <= 1) {
//            return s;
//        }
//
//        int n = s.length();
//        boolean[][] dp = new boolean[n][n];
//
//        int start = 0, end = 0;
//        int maxLength = 0;
//
//        for (int i = 0; i < n; i++) {
//            dp[i][i] = true;
//            start = i;
//            end = i;
//            maxLength = 1;
//        }
//
//        for (int i = 0; i < n - 1; i++) {
//            if (s.charAt(i) == s.charAt(i + 1)) {
//                dp[i][i + 1] = true;
//                start = i;
//                end = i + 1;
//                maxLength = 2;
//            }
//        }
//
//        for (int diff = 2; diff < n; diff++) {
//            for (int i = 0; i < n - diff; i++) {
//                int j = i + diff;
//                if (s.charAt(j) == s.charAt(i) && dp[i + 1][j - 1]) {
//                    dp[i][j] = true;
//
//                    if (j - i + 1 > maxLength) {
//                        start = i;
//                        end = j;
//                        maxLength = j - i + 1;
//                    }
//                }
//            }
//        }
//
//        return s.substring(start, end + 1);
//    }

}

File: LeetCode_Python/dp/strings/lps/LongestPalindromicSubstring.py
class LongestPalindromicSubstring:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        if n < 2:
            return s

        dp = [[False] * n for _ in range(n)]
        start, max_length = 0, 1

        for i in range(n):
            dp[i][i] = True

        for i in range(n - 1):
            if s[i] == s[i + 1]:
                dp[i][i + 1] = True
                start = i
                max_length = 2

        for length in range(3, n + 1):
            for i in range(n - length + 1):
                j = i + length - 1
                if s[i] == s[j] and dp[i + 1][j - 1]:
                    dp[i][j] = True
                    if length > max_length:
                        start = i
                        max_length = length

        return s[start: start + max_length]


if __name__ == '__main__':
    solution = LongestPalindromicSubstring()
    s1 = "babad"
    print("Expected Output: 'bab' or 'aba'")
    print("Actual Output:", solution.longestPalindrome(s1))

    s2 = "cbbd"
    print("Expected Output: 'bb'")
    print("Actual Output:", solution.longestPalindrome(s2))

    s3 = "a"
    print("Expected Output: 'a'")
    print("Actual Output:", solution.longestPalindrome(s3))
